lib/resources/shadow.rb

```diff
@@ -32,87 +30,55 @@ class Shadow < Inspec.resource(1)
+      @lines = @raw_content.to_s.split("\n")
       @filters = opts[:filters] || ''
+      raise Inspec::Exceptions::ResourceSkipped, 'The `shadow` resource is not supported on your OS.' unless inspec.os.unix?
       @params = @lines.map { |l| parse_shadow_line(l) }
     end
 
-    def filter(hm = {})
-      return self if hm.nil? || hm.empty?
-      res = @params
-      filters = ''
-      hm.each do |attr, condition|
-        condition = condition.to_s if condition.is_a? Integer
-        filters += " #{attr} = #{condition.inspect}"
-        res = res.find_all do |line|
-          case line[attr.to_s]
-          when condition
-            true
-          else
-            false
-          end
-        end
-      end
-      content = res.map { |x| x.values.join(':') }.join("\n")
-      Shadow.new(@path, content: content, filters: @filters + filters)
-    end
-
-    def entries
-      @lines.map do |line|
-        params = parse_shadow_line(line)
-        Shadow.new(@path, content: line,
-                   filters: "#{@filters} on entry user=#{params['user']}")
-      end
-    end
-
-    def users(name = nil)
-      name.nil? ? map_data('user') : filter(user: name)
-    end
-
-    def passwords(password = nil)
-      password.nil? ? map_data('password') : filter(password: password)
-    end
-
-    def last_changes(filter_by = nil)
-      filter_by.nil? ? map_data('last_change') : filter(last_change: filter_by)
-    end
-
-    def min_days(filter_by = nil)
-      filter_by.nil? ? map_data('min_days') : filter(min_days: filter_by)
-    end
-
-    def max_days(filter_by = nil)
-      filter_by.nil? ? map_data('max_days') : filter(max_days: filter_by)
-    end
-
-    def warn_days(filter_by = nil)
-      filter_by.nil? ? map_data('warn_days') : filter(warn_days: filter_by)
-    end
-
-    def inactive_days(filter_by = nil)
-      filter_by.nil? ? map_data('inactive_days') : filter(inactive_days: filter_by)
-    end
-
-    def expiry_dates(filter_by = nil)
-      filter_by.nil? ? map_data('expiry_date') : filter(expiry_date: filter_by)
-    end
+    filter = FilterTable.create
+    filter.add_accessor(:where)
+          .add_accessor(:entries)
+          .add(:user, field: 'user')
+          .add(:password, field: 'password')
+          .add(:last_change, field: 'last_change')
+          .add(:min_days, field: 'min_days')
+          .add(:max_days, field: 'max_days')
+          .add(:warn_days, field: 'warn_days')
+          .add(:inactive_days, field: 'inactive_days')
+          .add(:expiry_date, field: 'expiry_date')
+
+    filter.add(:content) { |t, _|
+      t.entries.map do |e|
+        [e.user, e.password, e.last_change, e.min_days, e.max_days, e.warn_days, e.inactive_days, e.expiry_date].compact.join(':')
+      end.join("\n")
+    }
+
+    filter.add(:users) { |u, _| u.entries.map(&:user) }
+
+    filter.add(:count) { |i, _|
+      i.entries.length
+    }
+
+    filter.connect(self, :params)
+
+    alias users user
```

Comment:
Should we add some deprecations here? Based on that improvement, I am not sure what you think users should use.

Reply:
Yes, I think we should deprecate the singular forms of user and password. I can update this PR with deprecation notices for those targeting 3.0. I'll update the docs too =)